### PR TITLE
Replace get_systems_within_jumps() from galaxy.py

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR replaces `get_systems_within_jumps()` with a server provided faster version based on the `Pathfinder` cache.

I mentioned this in PR #1225 about improving the galaxy generation.  The improvement without #1225 is marginal.

### CODE REVIEW

File: python/server/ServerWrapper.cpp [hunk 1]
```diff
@@ -1291,6 +1310,8 @@ namespace FreeOrionPython {
         def("create_field",                         CreateField);
         def("create_field_in_system",               CreateFieldInSystem);
 
+        def("systems_within_jumps",                 SystemsWithinJumps);
```
Comment (genestack-solomatin) [inline python/server/ServerWrapper.cpp]: You can add documentation string as third argument here.

File: python/server/ServerWrapper.cpp [hunk 2]
```diff
@@ -905,6 +905,25 @@ namespace {
...
+
+        std::vector<int> systems;
+
+        for (boost::python::stl_input_iterator<int> id(sys_ids); id != end; ++id) {
+            systems.push_back(*id);
+        }
+
+        auto near = GetPathfinder()->WithinJumps(jumps, systems);
+
+        for (auto ii : near) {
```
Comment (adrianbroher) [inline python/server/ServerWrapper.cpp]: ii?

File: python/server/ServerWrapper.cpp [hunk 3]
```diff
@@ -905,6 +905,25 @@ namespace {
...
+        list py_systems;
+        boost::python::stl_input_iterator<int> end;
+
+        std::vector<int> systems;
+
+        for (boost::python::stl_input_iterator<int> id(sys_ids); id != end; ++id) {
+            systems.push_back(*id);
+        }
+
+        auto near = GetPathfinder()->WithinJumps(jumps, systems);
```
Comment (adrianbroher) [inline python/server/ServerWrapper.cpp]: near?

File: universe/Pathfinder.cpp [hunk 4]
```diff
@@ -1010,6 +1015,51 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
 }
 
 
+void Pathfinder::PathfinderImpl::WithinJumpsCacheHit(
+    std::unordered_set<int>* result, size_t jump_limit,
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const {
+
+    // Scan the LUT of system ids and add any result from the row within
+    // the neighborhood range to the results.
+    for (auto it : m_system_id_to_graph_index) {
```
Comment (adrianbroher) [inline universe/Pathfinder.cpp]: it?

File: universe/Pathfinder.cpp [hunk 5]
```diff
@@ -1010,6 +1015,51 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
 }
 
 
+void Pathfinder::PathfinderImpl::WithinJumpsCacheHit(
+    std::unordered_set<int>* result, size_t jump_limit,
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const {
```
Comment (adrianbroher) [inline universe/Pathfinder.cpp:1020]: ii?

**Review by adrianbroher — CHANGES_REQUESTED** [review]
You should really reconsider your general variable naming style.

File: universe/Pathfinder.cpp (hunk 4 above)
Comment (geoffthemedio) [reply universe/Pathfinder.cpp]: not that unusual for an iterator type in FO code...

Comment (adrianbroher) [reply universe/Pathfinder.cpp]: True, but range based for loops don't use the iterator directly but the type that the iterator dereferences to.

**Review by Vezzra — APPROVED** [review]
Can't comment on the C++ side of things, as your implementation goes over my head, but moving this into the backend code is certainly a good idea.

Everything else looks good to me, builds and runs ok on OSX. LGTM (as far as I'm concerned).

**Comment by Cjkjvfnby** [comment]
> but moving this into the backend code is certainly a good idea
@Vezzra, why?

Python is slower when C++, but not so slow.  I expect that using cache should give more benefits then changing language. 

Code in python usually more readable.

File: python/server/ServerWrapper.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply python/server/ServerWrapper.cpp]: Thanks.

File: python/server/ServerWrapper.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply python/server/ServerWrapper.cpp]: Could you provide suggestions?

I've found that no matter what names I choose someone finds them too terse, too verbose, too cryptic or all of the above.

File: universe/Pathfinder.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply universe/Pathfinder.cpp:1020]: That is the row index, typically written  `i`.  However `ii` is much easier to grep.  This one I'd like to keep.

**Review by LGM-Doyle — COMMENTED** [review]
This is backend vs frontend issue.  

The backend can access all of the information to compute this result in a single vector.  The provided function `fo.systems_within_jumps(range, [systems])` is an intuitive  addition to the ServerWrapper.  The best place to do the calculation is where the data is.

File: python/server/ServerWrapper.cpp (hunk 3 above)
Comment (adrianbroher) [reply python/server/ServerWrapper.cpp]: After looking into the other changes why not use 'systems_in_vicinity'?  It was already a good name and the documentation for the python `get_systems_within_jumps` just provided the information I am missing out at `WithinJumps` so it should be transferred and adapted accordingly.

> I've found that no matter what names I choose someone finds them too terse, too verbose, too cryptic or all of the above.

That's the outcome of cooperative development. Everybody has to deal with different opinions. ;)  In genreal I will always complain about terse or generalized names if the context doesn't provide any information about the code usage. In this case `near` just didn't ring a bell.

File: python/server/ServerWrapper.cpp (hunk 2 above)
Comment (adrianbroher) [reply python/server/ServerWrapper.cpp]: `system_id` would probably better.

**Comment by Vezzra** [comment]
> Python is slower when C++, but not so slow.

IMO the speed advantage C++ offers is still significant enough to justify moving the implementation only for that (and then there are the reasons @LGM-Doyle cited).

> I expect that using cache should give more benefits then changing language.

Sure, but that's no reason not to take advantage of the additional performance boost the change to C++ offers.

> Code in python usually more readable.

Yes, but that's not really relevant in case of this function. Any speed gain, even if it's only a moderate one, is more important than the not quite as good readability.

**Review by LGM-Doyle — COMMENTED** [review]
I think that I have addressed the concerns.
I've also rebased it and used it to replace uses of get_systems_within_jumps() in the newly merged code from #1225.

**Comment by Vezzra** [comment]
Just noticed a minor thing: you did not remove the now obsolete `galaxy_topology.py`. Any reason for this?

**Review by adrianbroher — APPROVED** [review]

**Comment by LGM-Doyle** [comment]
Fixed.

The rebasing removed the original commit that removed the file.

**Pull request merged**